### PR TITLE
adding timestamp collector back from informer package

### PR DESCRIPTION
**Pull request by anvddriesch**
towards: https://github.com/giantswarm/giantswarm/issues/8077

### CODE REVIEW

**Comment by anvddriesch** [comment]
I tested this on `gauss`

File: controller/collector/spec.go [hunk 1]
```diff
@@ -0,0 +1 @@
+package collector
```
Comment (xh3b4sd) [inline controller/collector/spec.go]: We can delete the file if it does not serve any purpose. Usually we define interfaces here but there does not seem to be any. So dropping the spec file should be fine.

File: controller/controller.go [hunk 2]
```diff
@@ -207,6 +227,7 @@ func (c *Controller) bootWithError(ctx context.Context) error {
 		}
 
 		c.logger.LogCtx(ctx, "level", "debug", "message", "ensured custom resource definition exists")
+
```
Comment (xh3b4sd) [inline controller/controller.go]: Let's please not add too many empty lines where they do not serve a purpose. Especially when the pattern is not consistently applied.

File: controller/controller.go [hunk 3]
```diff
@@ -392,48 +413,33 @@ func (c *Controller) reconcile(ctx context.Context, req reconcile.Request) (reco
...
-		if err != nil {
-			return reconcile.Result{}, microerror.Mask(err)
-		}
 	}
 
 	if m.GetDeletionTimestamp() != nil {
 		event := "delete"
 
-		deletionTimestampGauge.WithLabelValues(t.GetKind(), m.GetName(), m.GetNamespace()).Set(float64(m.GetDeletionTimestamp().Unix()))
-		t := prometheus.NewTimer(eventHistogram.WithLabelValues(event))
```
Comment (xh3b4sd) [inline controller/controller.go]: This should stay. We only want to drop the deletion timestamp metric.

File: controller/controller.go [hunk 4]
```diff
@@ -392,48 +413,33 @@ func (c *Controller) reconcile(ctx context.Context, req reconcile.Request) (reco
...
 		c.logger.LogCtx(ctx, "level", "debug", "message", "reconciling object")
 		c.deleteFunc(ctx, obj)
 		c.logger.LogCtx(ctx, "level", "debug", "message", "reconciled object")
-
-		t.ObserveDuration()
 	} else {
 		event := "update"
 
-		creationTimestampGauge.WithLabelValues(t.GetKind(), m.GetName(), m.GetNamespace()).Set(float64(m.GetCreationTimestamp().Unix()))
-		t := prometheus.NewTimer(eventHistogram.WithLabelValues(event))
```
Comment (xh3b4sd) [inline controller/controller.go]: Same for creation. The timer should stay.

**Comment by anvddriesch** [comment]
thank you @xh3b4sd for the review! I reflected your comments in the new commit

File: controller/controller.go [hunk 5]
```diff
@@ -157,6 +172,7 @@ func (c *Controller) Boot(ctx context.Context) {
 	ctx = setLoggerCtxValue(ctx, loggerKeyController, c.name)
 
 	c.bootOnce.Do(func() {
+
```
Comment (xh3b4sd) [inline controller/controller.go]: Please not.

File: controller/collector/timestamp.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+func (t *Timestamp) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+	gvk, err := apiutil.GVKForObject(t.newRuntimeObjectFunc(), t.k8sClient.Scheme())
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(gvk)
+	list.SetKind(list.GetKind() + "List")
```
Comment (kopiczko) [inline controller/collector/timestamp.go]: That's a nit but I'd like to "finish" with `gvk` before starting with `list`. So'd prefer to put `gvk.Kind += "List"` in line 72 instead. I think this is less error-prone approach when it comes to code removal/replacement.

File: controller/collector/timestamp.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+			creationTimestampDesc,
+			prometheus.GaugeValue,
+			float64(object.GetCreationTimestamp().Unix()),
+			object.GetKind(),
+			object.GetName(),
+			object.GetNamespace(),
+		)
+
+		if object.GetDeletionTimestamp() != nil {
+
```
Comment (JosephSalisbury) [inline controller/collector/timestamp.go]:
```suggestion
```

File: controller/collector/timestamp.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	list.SetGroupVersionKind(gvk)
+	list.SetKind(list.GetKind() + "List")
+
+	err = t.k8sClient.CtrlClient().List(ctx, list)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, object := range list.Items {
+
```
Comment (JosephSalisbury) [inline controller/collector/timestamp.go]:
```suggestion
```

**Comment by anvddriesch** [comment]
Thank you for the reviews. :) I applied your suggestions. (Apparently I have a tendency to add useless blank lines, sorry about that)

**Comment by xh3b4sd** [comment]
When you have two approvals it is fair to go in case nobody requested more changes. When you want to get Joe's approval give him another shout. I don't think he meant to block here. Let's roll.

**Comment by JosephSalisbury** [comment]
Yeah, all good, didn't wanna block. I pretty much never request changes :D

**Pull request merged**